Decompiler recursive local function emits nonexistent helper call

## Product area

`ILInspector.Decompiler` product `member -S "Decompiled Source"` output for C# 7 local functions.

## False claim being falsified

Rung 4 claims local-function output is valid or honestly degraded. A recursive local function currently renders as a call to a nonexistent helper method while the selected method still appears as normal decompiled source. The output does not compile.

## Minimal repro

```csharp
namespace Rung4Probe;

public static class CSharp7Probe
{
    public static int RecursiveLocalFunction(int value)
    {
        int Fact(int n)
        {
            if (n <= 1)
            {
                return 1;
            }

            return n * Fact(n - 1);
        }

        return Fact(value);
    }
}
```

Build Release and inspect:

```bash
dotnet-inspect member CSharp7Probe --library Rung4Probe.dll -m RecursiveLocalFunction -S "Decompiled Source" --bare
```

## Observed bad output

```csharp
public static int RecursiveLocalFunction(int value) => CSharp7Probe.Fact(value);
```

Compiling the emitted shape fails:

```text
error CS0117: 'CSharp7Probe' does not contain a definition for 'Fact'
```

The real method is a compiler-generated local-function method with a metadata name such as `<RecursiveLocalFunction>g__Fact|...`; `Fact` is not an ordinary member on `CSharp7Probe`.

## First fix shape

Either recover recursive local functions with a local-function declaration that can call itself, or decline recovery and render an honest lowered/Partial form that does not invent a nonexistent `CSharp7Probe.Fact` member.

Do not treat the simple demangled name as an ordinary static method name unless a real spellable member exists.

## Done signal

- Recursive local-function repro no longer emits `CSharp7Probe.Fact(value)` at Full.
- Output is either valid recovered local-function syntax or visibly/honestly degraded.
- Existing nonrecursive local-function recovery in `LadderRung4GateTests` and `LocalFunctionRaisingPassTests` stays green.

Candidate blocker for #1599 rung 4.

--- a/src/ILInspector.Decompiler.Fixtures.LadderRung4/LadderRung4.cs
+++ b/src/ILInspector.Decompiler.Fixtures.LadderRung4/LadderRung4.cs
@@ -84,6 +84,21 @@ public static class CSharp7LocalSyntax
 
         return ref right;
     }
+
+    public static int RecursiveLocalFunction(int value)
+    {
+        return Fact(value);
+
+        static int Fact(int n)
+        {
+            if (n <= 1)
+            {
+                return 1;
+            }
+
+            return n * Fact(n - 1);
+        }
+    }
 }
 
 public readonly struct Point

--- a/src/ILInspector.Decompiler.Tests/LadderRung4GateTests.cs
+++ b/src/ILInspector.Decompiler.Tests/LadderRung4GateTests.cs
@@ -21,10 +21,12 @@ public class LadderRung4GateTests
         "<LocalFunctionCapturing>g__AddSeed|4_0",
         "<LocalFunctionCapturingWithLocal>g__AddSquare|5_0",
         "<LocalFunctionNonCapturing>g__Double|3_0",
+        "<RecursiveLocalFunction>g__Fact|9_0",
         "DeconstructPoint",
         "LocalFunctionCapturing",
         "LocalFunctionCapturingWithLocal",
         "LocalFunctionNonCapturing",
+        "RecursiveLocalFunction",
         "RefLocalUpdate",
         "SelectRef",
         "SimplePattern",
@@ -87,6 +89,13 @@ public class LadderRung4GateTests
         Assert.Equal(DecompilationFidelity.Partial, Fidelity("LocalFunctionCapturingWithLocal"));
         Assert.Contains("DisplayClass", localBodyCapture);
         Assert.Contains("AddSquare(3", localBodyCapture);
+
+        var recursiveLocal = Body("RecursiveLocalFunction");
+        Assert.Contains("return Fact(value);", recursiveLocal);
+        Assert.Contains("static int Fact(int n)", recursiveLocal);
+        Assert.Contains("return n * (Fact(n - 1));", recursiveLocal);
+        Assert.DoesNotContain("CSharp7LocalSyntax.Fact", recursiveLocal);
+        Assert.DoesNotContain("g__", recursiveLocal);
 
         var pattern = Body("SimplePattern");
         Assert.Contains("value is int", pattern);

--- a/src/ILInspector.Decompiler.Tests/LocalFunctionRaisingPassTests.cs
+++ b/src/ILInspector.Decompiler.Tests/LocalFunctionRaisingPassTests.cs
@@ -83,13 +83,15 @@ public class LocalFunctionRaisingPassTests
     }
 
     [Fact]
-    public void RecursiveLocalFunction_StaysLowered()
+    public void RecursiveLocalFunction_RaisesSelfCallableDeclaration()
     {
-        // The body calls itself; keeping the import non-recursive bars recovery.
         string output = PrintRaised(nameof(CfgSampleClass.RecursiveLocalFunction));
 
-        Assert.DoesNotContain("int Fact(int v) =>", output);     // no recovered declaration
-        Assert.Contains("Fact(n)", output);                      // call left as the lowered invocation
+        Assert.Contains("return Fact(n);", output);
+        Assert.Contains("static int Fact(int v)", output);
+        Assert.Contains("return v * (Fact(v - 1));", output);
+        Assert.DoesNotContain("CfgSampleClass.Fact", output);
+        Assert.DoesNotContain("g__", output);
     }
 
     [Fact]
@@ -211,6 +213,48 @@ public class LocalFunctionRaisingPassTests
 
         Assert.Equal(1, localImports);
         Assert.Equal(1, lambdaImports);
+        function.CheckInvariant();
+    }
+
+    [Fact]
+    public void LocalFunctionBodyCallingDifferentLocalFunction_StaysLowered()
+    {
+        var owner = TypeRef.Definition("Synthetic", "Samples", "Owner");
+        var first = new MethodRef(
+            owner,
+            "<M>g__First|0_0",
+            s_int,
+            [s_int],
+            HasThis: false)
+        {
+            CompilerGenerated = MetadataFactState.Yes,
+        };
+        var second = new MethodRef(
+            owner,
+            "<M>g__Second|0_1",
+            s_int,
+            [s_int],
+            HasThis: false)
+        {
+            CompilerGenerated = MetadataFactState.Yes,
+        };
+        var function = FunctionReturningCall(first, s_int);
+        var context = new PassContext(
+            new Stepper(enabled: false),
+            importMethodBody: method =>
+            {
+                if (method == first)
+                    return LocalFunctionBodyCalling(first, second);
+                if (method == second)
+                    return LocalFunctionBody(second, s_int);
+                return null;
+            });
+
+        new LocalFunctionRaisingPass().Run(function, context);
+
+        Assert.Empty(function.Descendants.OfType<LocalFunctionStatement>());
+        Assert.Empty(function.Descendants.OfType<LocalFunctionInvocation>());
+        Assert.Single(function.Descendants.OfType<Call>(), call => call.Callee == first);
         function.CheckInvariant();
     }
 
@@ -376,6 +420,23 @@ public class LocalFunctionRaisingPassTests
             isVirtual: false,
             new Constant(null, TypeRef.CoreLib("System", "Object")))));
         block.Add(new Return(new LoadArgument(0, "x", s_int)));
+        var body = new BlockContainer();
+        body.Add(block);
+        return new IrFunction(
+            localMethod.Name,
+            localMethod.DeclaringType,
+            new MethodSignature(s_int, [new Parameter("x", s_int)], HasThis: false, GenericParameterCount: 0),
+            [],
+            body);
+    }
+
+    static IrFunction LocalFunctionBodyCalling(MethodRef localMethod, MethodRef calledMethod)
+    {
+        var block = new Block();
+        block.Add(new Return(new Call(
+            calledMethod,
+            isVirtual: false,
+            [new LoadArgument(0, "x", s_int)])));
         var body = new BlockContainer();
         body.Add(block);
         return new IrFunction(

--- a/src/ILInspector.Decompiler.Tests/SubstratePredicateCensusTests.cs
+++ b/src/ILInspector.Decompiler.Tests/SubstratePredicateCensusTests.cs
@@ -18,6 +18,8 @@ public class SubstratePredicateCensusTests
             "Inlining-specific EH guard that first finds the containing catch clause before using ReferenceOwnership.IsInside on its filter.",
         [new("ExpressionInliningPass.cs", "SameRegions")] =
             "Inlining-specific EH range equality for candidate movement; the predicate is over function region spans, not IR identity.",
+        [new("LocalFunctionRaisingPass.cs", "SameLocalFunctionMethod")] =
+            "Local-function mangled names are unique within the declaring type; this pass-local check distinguishes self-recursion from nested or mutual local-function calls.",
         [new("NullCoalescingAssignmentPass.cs", "SameField")] =
             "FieldRef equality paired with pass-owned receiver/index-shape checks for null-coalescing assignment.",
         [new("NullCoalescingAssignmentPass.cs", "SameReceiver")] =

--- a/src/ILInspector.Decompiler/Pipeline/Passes/LocalFunctionRaisingPass.cs
+++ b/src/ILInspector.Decompiler/Pipeline/Passes/LocalFunctionRaisingPass.cs
@@ -67,13 +67,17 @@ public sealed class LocalFunctionRaisingPass : IIrPass
                 var body = importScope.Import();
                 if (body is null)
                     continue;
-                // Keep the import non-recursive: a body that calls a local function
-                // (itself, mutually, or nested) is out of this slice.
-                if (body.Descendants.OfType<Call>().Any(c => GeneratedCodeIdentity.IsLocalFunctionMethod(c.Callee)))
+                // Mutual or nested local-function calls are still out of this slice.
+                // A self-call is recoverable: rewrite it to the same local-function
+                // invocation used by the host call sites after the nested pipeline
+                // has run without re-entering this method's import.
+                if (HasOtherLocalFunctionCall(body, method))
                     continue;
 
                 IrPasses.Run(body, IrPasses.Default, context);
 
+                if (HasOtherLocalFunctionCall(body, method))
+                    continue;
                 if (environment is not null && !SubstituteEnvironment(body, environment))
                     continue;
                 bool allowLocals = environment is null;
@@ -83,6 +87,7 @@ public sealed class LocalFunctionRaisingPass : IIrPass
                     continue;
 
                 string name = CSharpNaming.MethodName(method.Name);
+                RewriteSelfCalls(body, method, name);
                 // The environment parameter is the last one; drop it from the source signature.
                 var parameters = environment is null
                     ? body.Signature.Parameters
@@ -243,6 +248,27 @@ public sealed class LocalFunctionRaisingPass : IIrPass
         return true;
     }
 
+    static bool HasOtherLocalFunctionCall(IrFunction body, MethodRef method)
+        => body.Descendants.OfType<Call>()
+            .Any(c => GeneratedCodeIdentity.IsLocalFunctionMethod(c.Callee)
+                && !SameLocalFunctionMethod(c.Callee, method));
+
+    static void RewriteSelfCalls(IrFunction body, MethodRef method, string name)
+    {
+        foreach (var call in body.Descendants.OfType<Call>().ToList())
+        {
+            if (!SameLocalFunctionMethod(call.Callee, method))
+                continue;
+
+            var arguments = call.DetachChildren().Cast<IrExpression>().ToList();
+            call.ReplaceWith(new LocalFunctionInvocation(name, method.ReturnType, arguments));
+        }
+    }
+
+    static bool SameLocalFunctionMethod(MethodRef left, MethodRef right)
+        => left.Name == right.Name
+            && Equals(left.DeclaringType, right.DeclaringType);
+
     static bool IsCaptureValue(IrExpression value, IrFunction function) => value switch
     {
         LoadArgument => true,
@@ -268,6 +294,8 @@ public sealed class LocalFunctionRaisingPass : IIrPass
             if (allowLocalStatements && statement is StoreLocal)
                 continue;
             if (allowLocalStatements && statement is StoreStackSlot)
+                continue;
+            if (statement is IfStatement)
                 continue;
             if (statement is not ExpressionStatement)
                 return false;

